### ISSUE TITLE
Fix autofix removing sourceMappingURL comments

### ISSUE THREAD

**Issue by flyerhzm** [stylelint/stylelint#3802]
> Clearly describe the bug

If the css file has only sourceMappingURL comment, cli with `--fix` option will raise an error.

> Which rule, if any, is the bug related to?

`no-empty-first-line`

> What CSS is needed to reproduce the bug?

```css


/*# sourceMappingURL=index.css.map */
```

> What stylelint configuration is needed to reproduce the bug?

e.g.

```json
{
  "rules": {
    "no-empty-first-line": "always"
  }
}
```

> Which version of stylelint are you using?

`9.8.0`

> How are you running stylelint: CLI, PostCSS plugin, Node API?

`stylelint "**/*.css" --config myconfig.json --fix`

> Does the bug relate to non-standard syntax (e.g. SCSS, Less etc.)?

No

> What did you expect to happen?

Do not raise error.

> What actually happened (e.g. what warnings or errors did you get)?

```shell
TypeError: Cannot read property 'raws' of undefined
 at /app/node_modules/stylelint/lib/rules/no-empty-first-line/index.js:37:25
```

**Comment by flyerhzm** [comment]
@jeddy3 I added a test case to reproduce this issue here https://github.com/xinminlabs/stylelint/commit/9768ff9431614d759028f2b7860189d38b681d29, it will generate the following error

```
  ● no-empty-first-line › reject › [true] › "\n/*# sourceMappingURL=index.css.map */\n" › empty first line with sourceMappingURL comment

    TypeError: Cannot read property 'raws' of undefined

      32 |       if (noEmptyFirstLineTest.test(rootString)) {
      33 |         if (context.fix) {
    > 34 |           root.nodes[0].raws.before = "";
         |                         ^
      35 |           return;
      36 |         }
      37 |

      at raws (lib/rules/no-empty-first-line/index.js:34:25)
      at cb (lib/utils/eachRoot.js:15:5)
      at eachRoot (lib/rules/no-empty-first-line/index.js:21:5)
      at Promise.resolve.then (lib/lintSource.js:218:14)
```

looks like it doesn't parse `sourceMappingURL` as a comment node, any suggestion how to fix it?

**Comment by jeddy3** [comment]
> I added a test case to reproduce this issue

Thanks. I can reproduce your error with a leading `/n`.

- `\n/*# sourceMappingURL=index.css.map */\n` throws an error with `--fix`
- `/*# sourceMappingURL=index.css.map */\n` is silently removed with `--fix`

Neither is good.

> any suggestion how to fix it?

I suggest we start by asking upstream in PostCSS what the intended behaviour is for these comments. If they are treated as normal comments, and [parsed as such](https://astexplorer.net/#/gist/139f70da1d2d8429f7aad0fa824b3047/e8b93f961d7e8fe6bc4b2196df0f2c17843b138e), then will need work out why stylelint isn't playing nicely with them. If they are treated as a special case by PostCSS then we need fix stylelint to account for that.

**Comment by kanlukasz** [comment]
Same problem here
I have a files that contains /*# sourceMappingURL=bootstrap.css.map */ in the last line of file.
Stylelint always deletes it. I would like to prevent this

**Comment by jeddy3** [comment]
Closing [as stylistic rules are frozen](https://github.com/stylelint/stylelint/issues/5674). The community is welcome to migrate the rule [to a plugin](https://stylelint.io/developer-guide/plugins) and fix this bug.

**Comment by jefferyto** [comment]
It seems to me that the removal of sourceMappingURL comments by `--fix` is a bug and unrelated to `no-empty-first-line` or any other stylistic rules. (The original issue description describes a situation involving `no-empty-first-line` but the deeper issue is the handling of sourceMappingURL comments in general.) This issue should be reopened.

**Comment by ybiquitous** [comment]
I just noticed that PostCSS seems to internally process a `# sourceMappingURL=` string in a comment:
https://github.com/postcss/postcss/blob/a0d9f103249a8db24d9f219ef23b05e75c78ee0a/lib/previous-map.js#L57

**Comment by besalraj30** [comment]
Hi

I’d like to work on this  issue. Could you assign it to me?

**Comment by Mouvedia** [comment]
I've labeled the issue as ready to implement. Please consider [contributing](https://stylelint.io/contributing) if you have time.

There are [steps on how to fix a bug in a rule](https://stylelint.io/developer-guide/rules#fix-a-bug-in-a-rule) in the Developer guide.

**Comment by romainmenke** [comment]
The root cause is that PostCSS strips source map comments when reading source files.
This is because PostCSS is primarily used in pipelines to bundle, transpile, ... CSS.

The original source map comment is read and removed. The original source map itself is taken into account when generating the final source map.

Because Stylelint doesn't generate source maps when fixing it effectively removes the source map comment because it doesn't add one of its own.

------

But it is more complicated than somehow preserving the original source map comment.

Consider this example:

```css
.empty-rule {}

* {
	background-color: oklch(50% 15% 140deg);
}

```

After some processing it produces:

```css
.empty-rule {}

* {
	background-color: rgb(81, 108, 75);
}

/*# sourceMappingURL=application.s2023-fall.min.css.map */

```

```json
{"version":3,"sources":["/wp-content/themes/validation/gulp/assets/css/application.css"],"names":[],"mappings":"AAAA,aAAa;;AAEb;CACC,kCAAuC;AACxC","file":"application.s2023-fall.min.css","sourcesContent":[".empty-rule {}\n\n* {\n\tbackground-color: oklch(50% 15% 140deg);\n}\n"]}
```

If we run autofix to remove empty rules we get:

```css
* {
	background-color: rgb(81, 108, 75);
}

/*# sourceMappingURL=application.s2023-fall.min.css.map */

```

If you then use that stylesheet and source map you will notice that it no longer works and that inspecting styles points to incorrect source locations.

The source map comment doesn't need to be preserved.
The entire source map needs to be read and a new source map needs to be produced.

-------

I think maybe these options are better:
- warn/error on source map comments and explain why generated files should not be linted. Users should instead lint the original source.
- override the original source map and append a new source map url

**Comment by Mouvedia** [comment]
> and a new source map needs to be produced

If it's simple enough that would be ideal.
I would reckon it's not really the responsibility of stylelint to regenerate a new source map.

---

I might have been too hasty to label it as _ready to implement_.
@besalraj30 were you planning to keep the bogus sourceMappingURL?
